File: Linked-List/Reorder-List_M.cpp
#include <iostream>
#include <deque>
using namespace std;

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    void reorderList(ListNode* head) {
        // Initialize an empty deque to store the nodes
        deque<ListNode*> mydeque;
        ListNode* temp = head;

        // Traverse the list and add each node to the deque
        while (temp != nullptr) {
            mydeque.push_back(temp);
            temp = temp->next;
        }

        int count = 1;
        ListNode dummy(0);  // Temporary dummy node to start the new list
        temp = &dummy;

        // Reorder the list by alternating between nodes from the front and back of the deque
        while (!mydeque.empty()) {
            ListNode* node;
            if (count % 2 == 1) {
                node = mydeque.front();  // Get a node from the front
                mydeque.pop_front();
            } else {
                node = mydeque.back();  // Get a node from the back
                mydeque.pop_back();
            }
            count++;
            temp->next = node;  // Link the current node to the new list
            temp = node;  // Move to the next node
        }

        temp->next = nullptr;  // Ensure the last node points to nullptr to terminate the list
    }
};

// Main function to demonstrate the reorderList function
int main() {
    // Creating a linked list: 1 -> 2 -> 3 -> 4
    ListNode* node4 = new ListNode(4);
    ListNode* node3 = new ListNode(3);
    node3->next = node4;
    ListNode* node2 = new ListNode(2);
    node2->next = node3;
    ListNode* head = new ListNode(1);
    head->next = node2;

    // Calling reorderList
    Solution().reorderList(head);

    // Printing the reordered list
    ListNode* temp = head;
    while (temp != nullptr) {
        cout << temp->val;
        if (temp->next != nullptr) {
            cout << " -> ";
        }
        temp = temp->next;
    }
    cout << endl;

    // Freeing allocated memory
//    delete node1;
    delete node2;
    delete node3;
    delete node4;

    return 0;
}

// Commented by GPT-4 on 19/7/2024

File: Linked-List/Reorder-List_M.py
from collections import deque
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        # Initialize an empty deque to store the nodes
        mydeque = deque()
        temp = head

        # Traverse the list and add each node to the deque
        while temp is not None:
            mydeque.append(temp)
            temp = temp.next

        count = 1
        temp = ListNode()  # Temporary dummy node to start the new list

        # Reorder the list by alternating between nodes from the front and back of the deque
        while mydeque:
            if count % 2 == 1:
                node = mydeque.popleft()  # Get a node from the front
            else:
                node = mydeque.pop()  # Get a node from the back
            count += 1
            temp.next = node  # Link the current node to the new list
            temp = node  # Move to the next node

        temp.next = None  # Ensure the last node points to None to terminate the list

# Example usage
if __name__ == "__main__":
    # Creating a linked list: 1 -> 2 -> 3 -> 4
    node4 = ListNode(4)
    node3 = ListNode(3, node4)
    node2 = ListNode(2, node3)
    head = ListNode(1, node2)

    # Calling reorderList
    Solution().reorderList(head)

    # Print the reordered list to verify
    temp = head
    while temp is not None:
        print(temp.val, end=" ->" if temp.next else "\n")
        temp = temp.next

# Commented by GPT-4 on 19/7/2024
